src/samples/mavlink/goto.c

```diff
@@ -0,0 +1,230 @@
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <sol-log.h>
+#include <sol-mainloop.h>
+#include <sol-mavlink.h>
+#include <sol-util.h>
+#include <stdio.h>
+
+#define TAKEOFF_ALT 10
+#define DEST_LAT -35.361354
+#define DEST_LONG 149.166218
+#define DEST_ALT 20
+
+// are we greater than altitude margin?
+#define GT_MARGIN(_lval, _rval)                 \
+    (_lval > (_rval * 0.95))                    \
+
+static void
+takeoff(struct sol_mavlink *mavlink)
+{
+    int err;
+    struct sol_mavlink_position takeoff = { 0 };
+
+    takeoff.altitude = TAKEOFF_ALT;
+    err = sol_mavlink_takeoff(mavlink, &takeoff);
+
+    if (err < 0) {
+        SOL_ERR("Could not takeoff: %s", sol_util_strerrora(-err));
+        return;
+    }
+
+    printf(">>>> Taking off.\n");
+}
+
+static void
+position_changed_cb(void *data, struct sol_mavlink *mavlink)
+{
+    int err;
+    struct sol_mavlink_position pos;
+
+    err = sol_mavlink_get_curr_position(mavlink, &pos);
+    if (err < 0) {
+        SOL_ERR("Could not get current position: %s", sol_util_strerrora(-err));
+        return;
+    }
+
+    if (sol_mavlink_check_armed(mavlink))
+        printf("lat: %f, long: %f, alt: %f\n", pos.latitude, pos.longitude,
+            pos.altitude);
+}
+
+static void
+mission_reached_cb(void *data, struct sol_mavlink *mavlink)
+{
+    int err;
+    sol_mavlink_mode mode;
+    struct sol_mavlink_position home, curr;
+    struct sol_mavlink_position dest = { DEST_LAT, DEST_LONG, DEST_ALT };
+
+    mode = sol_mavlink_get_mode(mavlink);
+    err = sol_mavlink_get_curr_position(mavlink, &curr);
+    if (err < 0) {
+        SOL_ERR("Could not get current position: %s", sol_util_strerrora(-err));
+        return;
+    }
+
+    if (mode != SOL_MAVLINK_MODE_GUIDED)
+        return;
+
+    if (GT_MARGIN(curr.altitude, DEST_ALT)) {
+        printf(">>>> Going back home.\n");
```

Comment:
it doesn't seem to be going back home.
I've tested it here:

>>>> Taking off.
lat: -35.363262, long: 149.165237, alt: 0.090000
[...]
lat: -35.363262, long: 149.165237, alt: 9.760000
>>>> Successful takeoff, starting a new mission, heading to: (-35.361355, 149.166214, 20.000000)
[...]
lat: -35.361382, long: 149.166199, alt: 19.120001
>>>> Going back home.
lat: -35.361359, long: 149.166214, alt: 19.059999
[...]
lat: -35.361351, long: 149.166214, alt: 0.090000
>>>> Landed...

Shouldn't it go back to lat: -35.363262, long: 149.165237 ?